traffic_monitor/datareq/crstate.go

```diff
@@ -84,10 +88,58 @@ func filterDirectlyPolledCaches(crstates tc.CRStates) tc.CRStates {
 	return filtered
 }
 
-func srvTRStateSelf(localStates peer.CRStatesThreadsafe, directlyPolledOnly bool) ([]byte, error) {
+func srvTRStateSelf(localStates peer.CRStatesThreadsafe, directlyPolledOnly bool, toData todata.TODataThreadsafe) ([]byte, error) {
 	if !directlyPolledOnly {
-		return tc.CRStatesMarshall(localStates.Get())
+		localStatesC := updateStatusSameIpServers(localStates, toData)
+		return tc.CRStatesMarshall(localStatesC)
 	}
-	unfiltered := localStates.Get()
+	unfiltered := updateStatusSameIpServers(localStates, toData)
 	return tc.CRStatesMarshall(filterDirectlyPolledCaches(unfiltered))
 }
+
+func updateStatusSameIpServers(localStates peer.CRStatesThreadsafe, toData todata.TODataThreadsafe) tc.CRStates {
+	localStatesC := localStates.Get()
+	toDataC := toData.Get()
+
+	for cache, _ := range localStatesC.Caches {
+		if _, ok := toDataC.SameIpServers[cache]; ok {
+			// all servers with same ip must be available if they are in reported state
+			allAvailableV4 := true
+			allAvailableV6 := true
+			allIsAvailable := true
+			for partner, _ := range toDataC.SameIpServers[cache] {
+				if partnerState, ok := localStatesC.Caches[partner]; ok {
+					// a partner host is reported but is marked down for too high traffic or load
+					// this host also needs to be marked down to divert all traffic for their
+					// common ip
+					if strings.Contains(partnerState.Status, string(tc.CacheStatusReported)) &&
+						strings.Contains(partnerState.Status, "too high") {
```

Comment:
I don't love searching human-readable message for a specific string, there's too many ways it could go wrong (what if human-readable part of the message changes? What if the message changes so there's still a match but the  meaning is different, like `"Good thing it's not too high"`? What if, for some reason, status messages get combined, so you get the first part without realizing it's invalid because there is more than 1 message in the concatenated string).

Much more reliable would be to store the result of those comparisons so we have a machine-readable way to know whether it's too high.